lambdas/build-cri-oauth-request/src/main/java/uk/gov/di/ipv/core/buildcrioauthrequest/helpers/SharedClaimsHelper.java

```diff
@@ -0,0 +1,94 @@
+
+public class SharedClaimsHelper {
+    private static final Logger LOGGER = LogManager.getLogger();
+    public static final String SHARED_CLAIM_ATTR_NAME = "name";
+    public static final String SHARED_CLAIM_ATTR_BIRTH_DATE = "birthDate";
+    public static final String SHARED_CLAIM_ATTR_ADDRESS = "address";
+    public static final String SHARED_CLAIM_ATTR_EMAIL = "emailAddress";
+    public static final String SHARED_CLAIM_ATTR_SOCIAL_SECURITY_RECORD = "socialSecurityRecord";
+
+    private SharedClaimsHelper() {}
+
+    public static SharedClaims generateSharedClaims(
+            String emailAddress, List<VerifiableCredential> vcs, List<String> allowedSharedClaims) {
+        var sharedClaims = new SharedClaims();
+
+        // Email address is provided separately in the session, not from a VC
+        sharedClaims.setEmailAddress(emailAddress);
+
+        // Other shared claims are added from successful VCs
+        for (var vc : vcs) {
+            if (VcHelper.isSuccessfulVc(vc)) {
+                var credentialSubject = vc.getCredential().getCredentialSubject();
+
+                if (credentialSubject instanceof PersonWithIdentity personWithIdentity
+                        && personWithIdentity.getName() != null) {
+                    sharedClaims.getName().addAll(personWithIdentity.getName());
+                }
+                if (credentialSubject instanceof PersonWithIdentity personWithIdentity
+                        && personWithIdentity.getBirthDate() != null) {
+                    sharedClaims.getBirthDate().addAll(personWithIdentity.getBirthDate());
+                }
+                if (ADDRESS.equals(vc.getCri())
+                        && credentialSubject instanceof AddressAssertion addressAssertion
+                        && addressAssertion.getAddress() != null) {
+                    sharedClaims.getAddress().addAll(addressAssertion.getAddress());
+                }
+                if (credentialSubject instanceof PersonWithDocuments personWithDocuments
+                        && personWithDocuments.getSocialSecurityRecord() != null) {
+                    sharedClaims
+                            .getSocialSecurityRecord()
+                            .addAll(personWithDocuments.getSocialSecurityRecord());
+                }
+            }
+        }
+
+        LOGGER.info(
+                LogHelper.buildLogMessage("Found shared claims")
+                        .with("names", sharedClaims.getName().size())
+                        .with("birthDates", sharedClaims.getBirthDate().size())
+                        .with("addresses", sharedClaims.getAddress().size())
+                        .with("emails", sharedClaims.getEmailAddress() == null ? 0 : 1)
+                        .with(
+                                "socialSecurityRecords",
+                                sharedClaims.getSocialSecurityRecord().size()));
+
+        stripDisallowedSharedClaims(sharedClaims, allowedSharedClaims);
```

Comment:
It might be safer to add this into the conditions above, but then you have to add a load more null-checking in order to distinguish 'present but empty' from 'not present'.

Now its in a separate class, it should be more obvious that you need to update this at the same time...